src/katgpucbf/xbgpu/kernels/beamform.mako

```diff
@@ -0,0 +1,120 @@
+/*******************************************************************************
+ * Copyright (c) 2023, National Research Foundation (SARAO)
+ *
+ * Licensed under the BSD 3-Clause License (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy
+ * of the License at
+ *
+ *   https://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+<%include file="/port.mako"/>
+
+#define N_BEAMS ${len(beam_pols)}
+#define BLOCK_TIME ${block_time}
+#define BLOCK_CHANNELS ${block_channels}
+#define QMAX 127
+
+// Quantise, and return saturation flag
+// TODO: this is copied from postproc.py. Create a shared implementation.
+DEVICE_FN bool quant(float value, int *out)
+{
+    float clamped = fminf(fmaxf(value, -QMAX), QMAX);
+    bool saturated = value != clamped;
+
+    int q;
+    // Convert to s8, round to nearest integer, and saturate
+    // (saturation is redundant due to the test above, but automatic
+    // for float-to-int conversions on CUDA hardware).
+#ifdef __OPENCL_VERSION__
+    q = convert_char_sat_rte(clamped);
+#else
+    asm("cvt.rni.sat.s8.f32 %0, %1;" : "=r" (q) : "f"(clamped));
```

Comment:
I'm gona need pointing towards some docs to make sense of this syntax 😅

Reply:
The actual "cvt.rni.sat.s8.f32" instruction is PTX; documentation [here](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cvt). The bits after the colon map the input and output registers to variables, and use the GCC conventions, documented [here](https://gcc.gnu.org/onlinedocs/gcc-13.2.0/gcc/Extended-Asm.html)